Fixups

haraldh opened the pull request:
A couple of fixups to the last commits, before any release is made.

haraldh committed:
fix: turn SnpPageType into enum with defined values

and remove the now redundant `value()` method.

haraldh committed:
fix: SnpStart: clarify comment about the migration userspace address

haraldh committed:
feat: split launch::launcher into launch::sev and launch::snp

This separates the two type-state machines clearly and avoids confusion.

npmccallum commented on src/launch/mod.rs:
```diff
@@ -350,37 +350,22 @@ bitflags! {
-        }
-    }
+    Cpuid = 0x6,
 }
```
This enum should probably have a `#[repr(..)]`.

haraldh replied:
fixed

npmccallum commented on src/launch/mod.rs:
```diff
@@ -348,39 +348,25 @@ bitflags! {
 /// Encoded page types for a launch update. See Table 58 of the SNP Firmware
 /// specification for further details.
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[repr(C)]
```
Also: `#[non_exhaustive]`.

Basically, all `enum`s should have `#[non_exhaustive]` unless there is very strong reason to avoid it.

haraldh replied:
right

npmccallum commented on src/launch/snp.rs:
```diff
@@ -0,0 +1,103 @@
+    pub fn as_mut_vmfd(&mut self) -> &mut U {
+        self.vm_fd
+    }
+}
```
Why not just `impl AsRawFd`? The `Launcher` itself is handling all the relevant ioctls on `/dev/sev`. So it doesn't make any sense to get the fd for `/dev/sev` out of the `Launcher` type. Right?

haraldh replied:
If I input a real kvm `VmFd`, I can use the methods of it then. Otherwise the VmFd is totally blocked by the Launcher holding a mutable reference.

npmccallum replied:
Then `impl AsRef` and `impl AsMut`.

haraldh replied:
yeah, maybe... I think I wanted to keep the door to the mut sev fd open

npmccallum replied:
I'm of the opposite opinion: nothing should have mut access to the sev fd while we're launching this guest.

haraldh replied:
Well, we need to create vcpus after SNP_LAUNCH_INIT and before SNP_LAUNCH_FINISH, and this can only be done with a VmFd.

npmccallum replied:
Yes, I agree. I'm not sure how that affects my point. Launch should `impl AsRef<VmFd>` but not `impl AsRef<SevFd>`.

haraldh replied:
Ah, right... so let's implement `impl AsRef<U>` and `impl AsMut<U>`

haraldh replied:
done

haraldh committed:
fix: silence clippy warnings

Tagged some Default impls with #[allow(clippy::derivable_impls)]
because clippy thinks the derive macro could do that, but it can't.

The derived version would not differ anyway, so what...

npmccallum commented on src/certs/sev/cert/v1/sig/mod.rs:
```diff
@@ -13,6 +13,7 @@ union Signatures {
     rsa: rsa::Signature,
 }
 
+#[allow(clippy::derivable_impls)]
```
This struct can derive `Default`.

haraldh replied:
```
error: this trait cannot be derived for unions
  --> src/certs/sev/cert/v1/sig/mod.rs:10:23
   |
10 | #[derive(Copy, Clone, Default)]
   |                       ^^^^^^^
   ```

npmccallum commented on src/kvm/types.rs:
```diff
@@ -120,6 +120,7 @@ pub struct SnpInit {
     flags: u64,
 }
 
+#[allow(clippy::derivable_impls)]
```
This struct can derive `Default`.

haraldh replied:
fixed
